src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,153 @@
+        }
+        elementData = Arrays.copyOf(elementData, newCapacity);
+    }
+    private static int hugeCapacity(int minCapacity) {
+        if (minCapacity < 0) {// overflow}
+            throw new OutOfMemoryError();
+        }
+        return (minCapacity > MAX_ARRAY_SIZE) ? Integer.MAX_VALUE
+                : MAX_ARRAY_SIZE;
+    }
+
     @Override
     public void add(T value) {
-
+        ensureCapacityInternal(size + 1);  // Increments modCount!!
+        elementData[size++] = value;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index > size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("illegal index!");
+        }
+        ensureCapacityInternal(size + 1);
+        System.arraycopy(elementData, index, elementData, index + 1,
+                size - index);
+        elementData[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        Object[] a = new Object[list.size()];
+        for (int i = 0; i < a.length; i++) {
+            a[i] = list.get(i);
+        }
+        int numNew = a.length;
+        ensureCapacityInternal(size + numNew);  // Increments modCount
+        System.arraycopy(a, 0, elementData, size, numNew);
+        size += numNew;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException("illegal index!");
+        }
+        return (T) elementData[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (index < 0 || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException("illegal index!");
+        }
+        elementData[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException("illegal index!");
+        }
+        Object oldValue = elementData[index];
+        int numMoved = size - index - 1;
+        if (numMoved > 0) {
+            System.arraycopy(elementData, index + 1, elementData, index,
+                    numMoved);
+        }
+        elementData[--size] = null; // clear to let GC do its work
+        return (T) oldValue;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        if (element == null) {
+            for (int index = 0; index < size; index++) {
+                if (elementData[index] == null) {
+                    fastRemove(index);
+                    return element;
+                }
+            }
+        } else {
+            for (int index = 0; index < size; index++) {
+                if (element.equals(elementData[index])) {
+                    fastRemove(index);
+                    return element;
+                }
+            }
+        }
+        throw new NoSuchElementException("Illegal value");
```

Comment:
The exception message 'Illegal value' could be more informative. Consider specifying what went wrong, such as 'Element not found in the list' .